recurring_contract/models/contract_group.py

```diff
@@ -304,47 +315,76 @@ def _calculate_start_date_and_offset(self):
             offset = 1
         return start_date, offset
 
-    def _should_skip_invoice_generation(self, invoicing_date):
+    def _should_skip_invoice_generation(self, invoicing_date, contract=None):
         """In such cases, we should skip the invoice generation:
         - There is already an invoice for this due date which has been cancelled or
           edited.
         - Contract group suspension.
+        - A specific contract is given and an invoice for this due date already exists
+          and isn't cancelled.
         """
         self.ensure_one()
-        dangling_invoices = self.env["account.move"].search(
-            [
-                "|",
+
+        if contract:
+            search_filter = [
+                "&",
                 "&",
+                "&",
+                "&",
+                "&",
+                ("state", "!=", "cancel"),
                 ("invoice_date_due", "=", invoicing_date),
                 ("partner_id", "=", self.partner_id.id),
+                ("move_type", "=", "out_invoice"),
+                ("line_ids.contract_id", "=", contract.id),
+                (
+                    "line_ids.product_id",
+                    "in",
+                    contract.product_ids.ids,
+                ),
+            ]
+        else:
+            search_filter = [
                 "&",
                 "&",
                 "&",
                 "&",
-                ("move_type", "=", "out_invoice"),
-                ("state", "=", "cancel"),
+                "&",
+                "|",
                 ("payment_state", "not in", ["paid", "not_paid"]),
+                ("state", "=", "cancel"),
+                ("invoice_date_due", "=", invoicing_date),
+                ("partner_id", "=", self.partner_id.id),
+                ("move_type", "=", "out_invoice"),
                 ("line_ids.contract_id", "in", self.active_contract_ids.ids),
                 (
                     "line_ids.product_id",
                     "in",
                     self.active_contract_ids.mapped("product_ids").ids,
                 ),
             ]
-        )
-        # Check for contract group suspension
-        is_suspended = (
-            self.invoice_suspended_until
-            and self.invoice_suspended_until > invoicing_date
-        )
 
-        return bool(dangling_invoices) or is_suspended
+        existing_invoices = self.env["account.move"].search(search_filter)
```

Comment:
`search_count` is faster and sufficient in this case.
```suggestion
        existing_invoices = self.env["account.move"].search_count(search_filter)
```

Reply:
Done.